### PR TITLE
kie-issues#1490: Use `turbo`'s diffing algorithm to know which packages are affected by changes to `pnpm-lock.yaml`, avoiding FULL builds

### LINKED ISSUE #1490: Use `turbo`'s diffing algorithm to know which packages are affected by changes to `pnpm-lock.yaml`, avoiding FULL builds
When using `turbo`'s filters, changing only `pnpm-lock.yaml` will tell us which packages are impacted, thus avoiding FULL builds on `kie-tools'`s PR checks. This is hugely beneficial, since adding/removing packages rarely requires a FULL build. The same is true for when bumping dependencies versions.

The code responsible for identifying the packages that changed based on a git diff is here:
https://github.com/apache/incubator-kie-tools/blob/main/.github/supporting-files/ci/build-partitioning/build_partitioning.ts#L153-L159

Since `turbo` would only be used to determine what needs to build, and not to build anything itself, we can use the same installation strategy we use for `bun` here:
https://github.com/apache/incubator-kie-tools/blob/main/.github/workflows/ci_build.yml#L79

It is expected that sending a PR changing only `pnpm-lock.yaml` won't require a FULL build for checking it, but rather a PARTIAL build for only the packages affected by the changes.

### PR DESCRIPTION
**Pull request by thiagoelg**
Closes https://github.com/apache/incubator-kie-issues/issues/1490

Unlike `pnpm`, `turbo` can detect what packages were affected by the changes to the `pnpm-lock.yaml` file. This is important because we can trigger the PARTIAL CI build instead of FULL builds when the `pnpm-lock.yaml` file is changed in a PR, making PR checks faster.

This PR also introduces the `PNPM_HOME` env var to the `devbox` setup, guaranteeing that `pnpm` uses the same store for local and global packages.

**NOTE.: When running `pnpm bootstrap` you may be asked to reinstall packages to the new store. Just type `y` and `pnpm` will deal with it.**

Here are some test PRs to validate the new diffing strategy with `turbo`:

- Changes to the lockfile only: https://github.com/thiagoelg/kie-tools/pull/39
- Adding a new package: https://github.com/thiagoelg/kie-tools/pull/40
- Simple change to a single package: https://github.com/thiagoelg/kie-tools/pull/38

### CODE REVIEW

File: .github/supporting-files/ci/build-partitioning/build_partitioning.ts [hunk 1]
```diff
@@ -183,10 +196,16 @@ async function getPartitions(): Promise<Array<None | Full | Partial>> {
         };
       }
 
-      const changedSourcePathsInPartition = changedSourcePaths.filter((path) =>
-        [...partition.dirs].some((partitionDir) => path.startsWith(`${partitionDir}/`))
+      console.log("[build-partitioning] Changed package dirs in Partition:");
```
Comment (tiagobento) [inline .github/supporting-files/ci/build-partitioning/build_partitioning.ts]: Please add the partition index here...Logs are very dense right now...

File: .github/supporting-files/ci/build-partitioning/build_partitioning.ts [hunk 2]
```diff
@@ -183,10 +196,16 @@ async function getPartitions(): Promise<Array<None | Full | Partial>> {
...
+      console.log(new Set(changedPackagesDirs));
+
+      const changedSourcePathsInPartition = changedPackagesDirs.filter((path) =>
+        [...partition.dirs].some((partitionDir) => path.startsWith(`${partitionDir}`))
       );
 
+      console.log("[build-partitioning] Changed source paths in Partition:");
```
Comment (tiagobento) [inline .github/supporting-files/ci/build-partitioning/build_partitioning.ts]: Please add the partition index here...Logs are very dense right now...

File: .github/supporting-files/ci/build-partitioning/build_partitioning.ts [hunk 3]
```diff
@@ -204,14 +223,13 @@ async function getPartitions(): Promise<Array<None | Full | Partial>> {
...
-        [...(await getDirsOfDependencies(affectedPackageNamesInPartition, partition.name))].map(
+        [...(await getDirsOfDependencies(affectedPackageNamesInPartition))].map(
           (pkgDir) => packageNamesByDir.get(pkgDir)!
         )
       );
 
-      console.log(`[build-partitioning] 'Partial' build of '${partition.name}'`);
       console.log(
-        `[build-partitioning] Building ${relevantPackageNamesInPartition.size}/${partition.dirs.size}/${allPackageDirs.size} packages.`
+        `[build-partitioning]: Building ${relevantPackageNamesInPartition.size}/${partition.dirs.size}/${allPackageDirs.size} packages.`
```
Comment (tiagobento) [inline .github/supporting-files/ci/build-partitioning/build_partitioning.ts]:
```suggestion
        `[build-partitioning] Building ${relevantPackageNamesInPartition.size}/${partition.dirs.size}/${allPackageDirs.size} packages.`
```

File: .github/supporting-files/ci/build-partitioning/globals.ts [hunk 4]
```diff
@@ -22,7 +22,7 @@ import * as fs from "fs";
 
 export const __ROOT_PKG_NAME = "kie-tools-root";
 
-export const __PACKAGES_ROOT_DIRS = ["packages", "examples"];
+export const __PACKAGES_ROOT_PATHS = ["packages/", "examples/", "repo/", "pnpm-lock.yaml"];
```
Comment (tiagobento) [inline .github/supporting-files/ci/build-partitioning/globals.ts:33]: Can we add a TSDoc comment here explaining what these paths are?

File: .github/workflows/ci_build.yml [hunk 5]
```diff
@@ -87,7 +91,7 @@ jobs:
             --partition='./.github/supporting-files/ci/partitions/partition0.txt' \
             --partition='./.github/supporting-files/ci/partitions/partition1.txt'
 
-          npm -g uninstall bun
+          npm -g uninstall bun turbo
```
Comment (tiagobento) [inline .github/workflows/ci_build.yml:94]: I'd say deleting turbo.json here would be a good cleanup as well.

File: .github/supporting-files/ci/build-partitioning/build_partitioning.ts [hunk 6]
```diff
@@ -240,8 +258,12 @@ async function getPartitions(): Promise<Array<None | Full | Partial>> {
...
-      .map((pkgDir) => pkgDir.split(path.sep).join(path.posix.sep))
+    stdoutArray(execSync(`bash -c "pnpm ${packagesFilter} exec bash -c pwd"`).toString()) //
+      .map((pkgDir) => convertToPosixStylePath(pkgDir))
   );
 }
+
+function convertToPosixStylePath(pathString: string) {
+  const alreadyRelativePath = !pathString.startsWith(path.sep) && !pathString.startsWith(path.posix.sep);
+  return `${alreadyRelativePath ? pathString : path.relative(cwd, pathString)}`.split(path.sep).join(path.posix.sep);
+}
```
Comment (tiagobento) [inline .github/supporting-files/ci/build-partitioning/build_partitioning.ts]: So this function's name doesn't really help us understand what it returns. It gets a "pathString", and it returns a "posixStylePath". It doesn't say much about whether both paths are relative or absolute, and whether the returned path is relative or absolute. I've been using the following pattern and it helped me a lot --> https://github.com/apache/incubator-kie-issues/issues/783#issuecomment-1871525975

**Review by tiagobento — CHANGES_REQUESTED** [review]
Thanks @thiagoelg. Great improvement!!! Left some minor comments inline.

File: .github/supporting-files/ci/build-partitioning/build_partitioning.ts (hunk 1 above)
Comment (thiagoelg) [reply .github/supporting-files/ci/build-partitioning/build_partitioning.ts]: I can probably remove this log

File: .github/supporting-files/ci/build-partitioning/build_partitioning.ts (hunk 2 above)
Comment (thiagoelg) [reply .github/supporting-files/ci/build-partitioning/build_partitioning.ts]: This one too

**Review by tiagobento — APPROVED** [review]

**Review by porcelli — APPROVED** [review]
yay!

**Pull request merged**